kernel/src/table_changes/data_read.rs

```diff
@@ -0,0 +1,217 @@
+use std::iter;
+
+use itertools::Either;
+use roaring::RoaringTreemap;
+use url::Url;
+
+use super::scan_file::{CdfScanFileType, ResolvedCdfScanFile, UnresolvedCdfScanFile};
+use crate::actions::deletion_vector::selection_treemap_to_bools;
+use crate::table_changes::scan_file::CdfScanFile;
+use crate::{DeltaResult, Engine, Error};
+
+/// Resolves the deletion vectors for an [`UnresolvedCdfScanFile`]. This function handles two
+/// types of `CdfScanFile`s:
+/// 1. The first case is a [`CdfScanFile`] paired with a remove deletion vector. The `scan_type`
+///    must be [`CdfScanFileType::Add`]. In this case, both the add and remove deletion vectors are
+///    read if they exist. Then, we find the set of rows in the scan file that have been added, and
+///    the set of rows that have been removed. The set of removed rows (if any) will be represented
+///    by a [`ResolvedCdfScanFile`] with `scan_type` = [`CdfScanFileType::Remove`]. The set of
+///    added rows (if any) will be represented by a [`ResolvedCdfScanFile`] with `scan_type` =
+///    [`CdfScanFileType::Add`].
+///
+///    Note: We allow the possibility for there to be both added rows, and removed rows for a
+///    single add/remove pair.
+/// 2. The second case handles all other add, remove, and cdc [`CdfScanFile`]s. These will simply
+///    have the deletion vector read (if present), and each is converted into a [`ResolvedCdfScanFile`].
+///    No changes are made to the `scan_type`.
+#[allow(unused)]
+pub(crate) fn resolve_scan_file_dv(
+    engine: &dyn Engine,
+    table_root: &Url,
+    scan_file: UnresolvedCdfScanFile,
+) -> DeltaResult<impl Iterator<Item = ResolvedCdfScanFile>> {
+    let UnresolvedCdfScanFile {
+        scan_file,
+        remove_dvs,
+    } = scan_file;
+    let paired_rm_dv = remove_dvs.get(&scan_file.path);
+    match (&scan_file.scan_type, paired_rm_dv) {
+        (CdfScanFileType::Add, Some(rm_dv)) => {
+            // Helper function to convert a treemap to a [`ResolvedCdfScanFile`]. The `scan_type`
+            // of the [`ResolvedCdfScanFile`] is set to `out_type` This returns an empty iterator
+            // if nothing is selected.
+            fn treemap_to_iter(
```

Comment:
We map scan file to iter<resolved scan file> because we could have 0 (filtered out) 1 (normal) or 2 (adds+removes) results? Which in turn requires _two_ `Either` so we can capture four possible cases? I wonder if it would be simpler to always produce the iterators and let the caller flatten() them? 

Something like:
```rust
let add_dv = scan_file.add_dv.dv_info.get_treemap(...)?:
let rm_dv = scan_file.rm_dv.dv_info.get_treemap(...)?:
let (add_dv, rm_dv) = match (scan_file.add_dv, scan_file.rm_dv, scan_file.scan_type) {
    (Some(_), _, CDFScanFileType::Remove) => {
        return Err(... remove cannot have add DV ...);
    }
    (_, Some(_), CDFScanFileType::Cdc) => {
        return Err(... cdc cannot have remove DV ...);
    }
    (Some(add_dv), Some(rm_dv), _) => {
        // Take the symmetric difference so we don't double count rows
        (Some(&add_dv - &rm_dv),  Some(&rm_dv - &add_dv))
    }
    (add_dv, None, CDFScanFileType::Add | CDFScanFileType::Cdc) => {
        (Some(add_dv.get_or_else(|| Default::default())), None)
    }
    (None, rm_dv, CDFScanFileType::Remove) => {
        (None, Some(rm_dv.get_or_else(|| Default::default())))
    }
};

let adds = add_dv.map(treemap_to_bools)
    .map(|sel| ResolvedCDFScanFile { ... });
let removes = rm_dv.map(treemap_to_bools)
    .map(|sel| sel.into_iter().map(Not::not).collect())
    .map(|sel| ResolvedCDFScanFile { ... });
[adds, removes].into_iter().flatten()
```

Reply:
... except I think I got the selection vector inversion bit wrong; it should only be inverted if it was a remove DV on an Add action.

Reply:
Regarding `treemap_to_bools`: I noticed that it initializes the vec with all `true` be default, then sets elements to `false`. I observed that when I initialized the vec with all `false` and only selected elements to `true`, it's significantly faster. I found a post explaining [why this may be](https://stackoverflow.com/questions/71946484/why-is-vec0-super-large-number-so-memory-efficient-when-the-default-value). Zeroed out pages seem to work well with the kernel :) 

Given that the selection vector is `O(#rows)` it may worth avoiding the overhead of iterating over it again with `Not::not` and take advantage of  the zeroed out allocation. 

I'm assuming here `#rows` could be very large.

Reply:
I came across a weird case 🤔. The way you have it, we seem to have 
```rust
CdfScanFile {
    add_dv: DvInfo,
    rm_dv: DvInfo
    ...
}
```
where the deletion vector in add_dv/rm_dv could be `None`.

The problem comes in when differentiating the following cases:
1. `[Remove(path=A, dv=None), Add(path=A, dv=X)]`
2. `[Add(path=A, dv=X)]`.

This has got me thinking whether case 2 would ever happen.  Suppose we want to handle it:

I think both of these would fall into this case:
```rust
(add_dv, None, CDFScanFileType::Add | CDFScanFileType::Cdc) => {
        (Some(add_dv.get_or_else(|| Default::default())), None)
}
````

So I'm instead electing to go with something like this:
```rust 
CdfScanFile {
    deletion_vector: DvInfo,
    remove_dv: Option<DvInfo>
    ...
}
```

so that the cases become clear:
1. `remove_dv = Some(DvInfo( deletion_vector: None))`
2. `remove_dv = None`

Reply:
Yes, I can totally imagine zero-initialized is more efficient for many reasons (tho correctness trumps performance, if there is friction between the two).

Your newer approach to handling add vs. rm DV makes a lot of sense, thanks!